javascript/57-Insert-Interval.js

```diff
@@ -1,26 +1,42 @@
-let insert = function(intervals, newInterval) {
-    let res = [];
-    let isAdded = false;
-
-    intervals.forEach(int => {
-        if (int[0] > newInterval[1]) {
-            if (!isAdded) {
-                res.push(newInterval);
-                isAdded = true;
-            }
-
-            res.push(int);
-        } else if (int[1] < newInterval[0]) {
-            res.push(int);
-        } else {
-            newInterval[0] = Math.min(newInterval[0], int[0]);
-            newInterval[1] = Math.max(newInterval[1], int[1]);
-        }
-    });
-
-    if (!isAdded) {
-        res.push(newInterval);
-    }
+/**
+ * @param {number[][]} intervals
+ * @param {number[]} newInterval
+ * Time O(N) | Space O(N)
+ * @return {number[][]}
+ */
+ var insert = function(intervals, newInterval) {
+    const { beforeIndex, before } = getBefore(intervals, newInterval);
+    const afterIndex = mergeIntervals(intervals, newInterval, beforeIndex);
+    const after = intervals.slice(afterIndex);
 
-    return res;
+    return [ ...before, newInterval, ...after ];
 };
+
+const getBefore = (intervals, newInterval, index = 0, before = []) => {
+    const hasGap = ([ prevStart, prevEnd ], [ currStart, currEnd ]) =>
```

Comment:
same as others

Reply:
@mitchellirvin 
Since this in a while loop, we can either leave as is

```js
const hasOverlap = () => ...
while (hasOverlap())
```

Move out of scoop

```js
const hasOverlap = () => ...

const loop = () => {
     while (hasOverlap())
}
```

or add inline with comment

```js
// hasOverlap = intervals[index] <= newInterval[1]
while (intervals[index] <= newInterval[1])
```